### DISCUSSION TITLE
WebAssembly configuration

### DISCUSSION THREAD

**Discussion by artmedia** [oqtane/oqtane.framework#790] [Q&A] [answered]
Hello,
I'm trying to configure default site (localhost) from Server to WebAssembly.
In /admin/site page, in section "Progressive Web Application Settings" I changed the option "Is Enabled?" to yes.
Aftrer that, I get this error:  "Error: System.InvalidOperationException: Nullable object must have a value. at System.Nullable`1.get_Value() at Oqtane.UI.ThemeBuilder.InitializePwa(Interop interop) in D:\mySiti2\oqtane.framework-master\Oqtane.Client\UI\ThemeBuilder.razor:line 65 at Oqtane.UI.ThemeBuilder.OnParametersSetAsync() in D:\mySiti2\oqtane.framework-master\Oqtane.Client\UI\ThemeBuilder.razor:line 51 at Microsoft.AspNetCore.Components.ComponentBase.CallStateHasChangedOnAsyncCompletion(Task task)
 /_framework/blazor.server.js:19_"

I also tryed to modify appsettings.json "Runtime" from "Server" to "WebAssembly" but nothing happens (still get the error).

I'm using v.1.0.4, is there a tutorial for WASM configuration?
thnaks

**Comment by hishamco** [comment]
> I'm using v.1.0.4, is there a tutorial for WASM configuration?

Nothing but set the `Runtime` as you did from `Server` to `Client`

Regarding your exception I will try to have a look and fix the issue if there's a bug

**Comment by artmedia** [comment]
if can help, after modifying appsettings.json "Runtime" from "Server" to "WebAssembly", while debugging, I noticed that line 42 in _Host.cshtml ('@if Configuration.GetSection("Runtime").Value == "WebAssembly"') "value" still contintinues to be "Server"

**Reply by hishamco** [reply]
I don't think so ;)  try to rebuild and make sure the `Oqtane.Server` is set to startup project

**Comment by artmedia** [comment]
I downloaded latest version (1.0.4) few minutes ago.
Unzipped it and opened in VS2019 (16.7.5) and launched debug mode of Server project in IIS Express.
I see homepage in Blazor Server.
Then:

1) modify appsettings.json "Runtime" from "Server" to "WebAssembly" and launch again.
Then I see homepage but It doesn't seem to be Blazor WebAssembly. 
I see reference to _framework/blazor.webassembly.js but I do NOT see reference to /manifest.json.
In browser (I use Edge) I should see an icon asking if I want to install app (see image) 
![screen](https://user-images.githubusercontent.com/4210561/95673615-7d54fe00-0baa-11eb-9b05-263bd86a0069.png)

2) In Admin dashboard, Site setting, in section "Progressive Web Application Settings" if I change option "Is Enabled?" to yes, stop IIS Express and launch it again, then it CRASHES.
(i have to manually edit field PwaIsEnabled in Site table)
What's the use of this option?

**Reply by hishamco** [reply]
This will allow you to create a Progressive Web App, it will act as in app in your mobile and PC too if I am not wrong, so no need to open it from the browser

**Reply by artmedia** [reply]
@hishamco thanks for reply, I suggest to create an Hello World project of PWA type in VS, it takes just few minutes. You can see how it works in browser. I'm not an expert but I think all PWA app should behave in the same way. 
If not all PWA app behave in the same way, what is the procedure to install Oqtane PWA?

**Reply by hishamco** [reply]
> what is the procedure to install Oqtane PWA?

Frankly I didn't try it with Oqtane but may be there's an admin feature to enable it

**Reply by artmedia** [reply]
perhaps @sbwalker can answer this

**Reply by hishamco** [reply]
Let me have a look now ...

**Reply by hishamco** [reply]
@artmedia I'm successfully set the PWA from the admin nothing but the icon you show before in your screenshot seems not appear, after I open the browser console I see the following error

![PWA](https://user-images.githubusercontent.com/3237266/95676227-33c5dc80-0bc5-11eb-89ec-07f7674aac47.png)

@sbwalker any idea?

**Reply by artmedia** [reply]
it seems to me that manifest.json is missing

**Comment by sbwalker** [comment]
Just to clarify, WebAssembly and Progressive Web Applications (PWA) have nothing to do with one another - they are completely different technology concepts. 

WebAssembly is a new type of code that can be run in modern web browsers — it is a low-level assembly-like language with a compact binary format that runs with near-native performance. Blazor WebAssembly allows you to write full-stack web apps in C# which can run natively in a web browser. 

A PWA is an app built with standard web technologies that provides functionality that rivals an actual native mobile app. Using a "service worker" you can turn almost any website into a PWA. 

Microsoft included PWA support in its most recent Blazor WebAssembly template - which may have created your confusion. The reality is that the Blazor Server template could also be easily enhanced to provide PWA features as well. 

If you want to run Oqtane as a Blazor WebAssembly app, you simply need to specify "WebAssembly" for the Runtime setting in appsettings.json. When you run the application it will include blazor.webassembly.js and download the application to the browser on startup.

If you want to run Oqtan on Blazor Server, you simply need to specify "Server" or "" for the Runtime setting in appsettings.json. When you run the application it will include blazor.server.js on startup.

If you want to enable PWA functionality when running Oqtane on either Blazor WebAssembly or Blazor Server ( both hosting options are supported ) you need to login to the Site and go to Site Settings in the Admin Dashboard and enable the Progressive Web Application option. You also need to provide an App icon and a Splash icon in the required sizes ( these are mandatory based on the PWA specification ). Once you have provided these options, the framework will dynamically generate the manifest.json and include a link to the service worker ( see Oqtane.Client\UI\ThemeBuilder.razor ). It generates these scripts dynamically because Oqtane is a framework which supports multi-tenancy and it needs to use the unique options set for each site. If you are experiencing errors it is probably because you did not provide an app icon and splash icon.

**Reply by artmedia** [reply]
Shaun,
Oqtane architecture is powerful and flexible at the same time, for this reason I like it very much and I'm planning to use it for my customer's projects.
As I can see the use of Progressive Web Application will be increasing, so I need to be test this feature of Octane.
After setting Progressive Web Application option in Site Settings and after providing an App icon and a Splash icon I still got a js error.
My procedure is this:
1) fresh Oqtane 1.0.4 install
2) modified appsettings.json "Runtime" value from "Server" to "WebAssembly"
3) login as Host, in Site Settings set Progressive Web Application option and provided App icon and Splash icon (uploaded 2 png respectively 192x192 and 512x512)
After restarting Oqtane in IIS Express, I got javascript error: "SyntaxError: Unexpected token 'const'" as @hishamco showed and no PWA installable icon

For Oqtane PWA should I see (in Edge Chromium) the installable icon as shown here?  
https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/Add_to_home_screen

**Reply by hishamco** [reply]
@sbwalker could you please check the PWA feature, I already test it with the latest bits, it's not working as expected

**Reply by sbwalker** [reply]
Yes, Oqtane should provide standard PWA functionality when the feature is enabled. I will take a look to try and identify the problem you are experiencing.

**Reply by sbwalker** [reply]
I am guessing it is because line 82 in ThemeBuilder.razor does not have a semicolon after the object literal declaration. On line 83 it then declares a const which probably throws the error. In browsers such as Firefox ( which is what I use ) it does not throw an error - it considers the semi-colon to be optional in this case. However I have read that IE/Edge consider the semi-colon to be mandatory. I will submit a PR.

**Reply by artmedia** [reply]
great, 
thanks

**Reply by hishamco** [reply]
Nice catch!!

**Comment by sbwalker** [comment, marked as answer]
fixed in #817
